README.md
[![logo](docs/logo.png)](docs/home.md "for developers")  

Module of Tools.  
Manipulating types.  
---

1) [Таблица](docs/table/table.md)  
2) [История](docs/history.md)  




docs/history.md
История tools-types 
-------
Основан: 20 февраля 2021 15:19:37, Картонажником.  
[2021y-02m-20d][15:19:37] Idrisov Denis R.  
[2021y-02m-20d][15:19:37] birthday of the project  


**История изменений**  

**Версия 1.2.7 (предварительная)**  
----------------------------------
[2021y-04m-25d][23:50:00] #74_v127
1) upd: types/sfinae/deref_available   100 PRE
     - upd: test-deref_available

**Версия 1.2.7 (предварительная)**  
----------------------------------
[2021y-04m-24d][23:50:00] #74_v127
1) bug: types/sfinae/sfinae-2010  100 PRE
2) add: types/sfinae/sfinae-2008  100 PRE
     - deref_available          100 PRE

**Версия 1.2.7 (предварительная)**  
----------------------------------
[2021y-04m-23d][23:55:00] #72_v127
1) add: types/sfinae 100 PRE
     add: sfinae-2015  100 PRE (ready)
     add: sfinae-2010  100 PRE (not worked)
     add: sfinae-2008  100 PRE (in progress...)

**Версия 1.2.7 (промежуточная)**  
----------------------------------
[2021y-04m-23d][13:15:50] #71_docs
1) fix docs errors

**Версия 1.2.7 (промежуточная)**  
----------------------------------
[2021y-04m-22d][12:00:00] #69_docs
1) обновлено: тесты traits
2) баг-фикс: array-2008      102 PRE

**Версия 1.2.7 (промежуточная)**  
----------------------------------
[2021y-04m-21d][11:00:00] #67_docs
1) обновлен: traits                     013 PRE
     - добавлен: 
       - is_convertible      001 PRE
       - is_lvalue_reference 001 PRE
       - is_rvalue_reference 001 PRE

2) добавлен: sfinae                     100 PRE
     - добавлен: has_reserve     100 PRE
     - добавлен: has_splice      100 PRE

3) обновлен: common                     115 PRE
     - обновлен: array                102 PRE
       - обновлен: array-2008       102 PRE
       - обновлен: array-2012       102 PRE
       - обновлен: array-2013       102 PRE
     - обновлен: is_volatile_data     102 PRE
       - обновлен: is_volatile_data-2013  102 PRE
     - обновлен: is_const_data     102 PRE
       - обновлен: is_const_data-2013  102 PRE
     - перемещены в sfinae:
       - is_dereferencable       102 PRE
       - is_lambda               101 PRE
       - is_incrementable        101 PRE
       - is_decrementable        101 PRE
       - is_iterator             101 PRE
       - is_iterable             101 PRE
       - has_value_type          101 PRE
       - has_mapped_type         101 PRE
       - has_operator_access     101 PRE
       - has_begin               101 PRE
       - has_end                 101 PRE

**Версия 1.2.7 (промежуточная)**  
----------------------------------
[2021y-04m-20d][16:11:27] #65_docs
1) обновлен: pragma                     103 PRE
     - добавлено версионирование
2) обновлен: macro                      101 PRE
     - добавлено версионирование

3) исправлен: types/variadic          003 PRE
     - исправлен: variadic-old      003 PRE
     - обновлен:  variadic-new      003 PRE
     - обновлены: тесты

**Версия 1.2.7 (промежуточная)**  
----------------------------------
[2021y-04m-19d][11:54:01] #63_docs
1) обновлен: version                    127 PRE
2) обновлен: types                      127 PRE
  2.1) обновлен void_t.hpp          101 PRE
    - [исправлен баг](articles\using-declaration.md)
    - [исправлена проблема](articles\void_t.md)
    - обновлены: тесты

  2.2) разделяем fixed на fixed и limit:
    - обновлен: fixed               103 PRE
      - обновлены: тесты
    - добавлен: limit               100 PRE
      - добавлены: тесты

**Версия 1.2.6 (предварительная)**  
----------------------------------
[2021y-04m-14d][01:40:23] #61_is_iterable

1) обновлен: version                    126
2) обновлен: types                      126
3) обновлен: common                     114

  3.1) добавлен: is_incrementable     100
       - is_incrementable-2008.hpp  100
       - is_incrementable-2012.hpp  100
       - is_incrementable-2013.hpp  100

  3.1.1) добавлен: is_prefix_incrementable
  3.1.2) добавлен: is_postfix_incrementable 
  3.1.3) добавлен: is_incrementable

  3.2) добавлен: is_decrementable     100
       - is_decrementable-2008.hpp  100
       - is_decrementable-2012.hpp  100
       - is_decrementable-2013.hpp  100

  3.2.1) добавлен: is_prefix_decrementable
  3.2.2) добавлен: is_postfix_decrementable 
  3.2.3) добавлен: is_decrementable

  3.4) добавлен: is_iterator          100
       - is_iterator-2008.hpp       100
       - is_iterator-2012.hpp       100
       - is_iterator-2013.hpp       100
       - is_iterator-2015.hpp       100

  3.5)  добавлен: is_iterable         100
  3.6)  добавлен: has_value_type      100
  3.7)  добавлен: has_mapped_type     100
  3.8)  добавлен: has_operator_access 100
  3.9)  добавлен: has_begin           100 PRE
  3.10) добавлен: has_end             100 PRE

  ~~is_container~~
  ~~is_linear_containr~~
  ~~is_mapped_container~~
  ~~is_realloc_containr~~

4) добавлен: types/void_t.hpp 100 PRE
     - void_t
     - dSFINAE_CHECK
     - dSFINAE_EXPRESSION
     - dSFINAE_DECLARATION_1
     - dSFINAE_DECLARATION_2

5) добавлен: test-200-types/types/common/is_incrementable
     - добавлен: test-is_postfix_incrementable
     - добавлен: test-is_prefix_incrementable
     - добавлен: test-is_incrementable
6) добавлен: test-200-types/types/common/is_decrementable
     - добавлен: test-is_postfix_decrementable
     - добавлен: test-is_prefix_decrementable
     - добавлен: test-is_decrementable

7) добавлены тесты:
     - добавлен: test-is_iterator
     - добавлен: test-is_iterable
     - добавлен: test-has_value_type
     - добавлен: test-has_mapped_type
     - добавлен: test-has_operator_access
     - добавлен: test-has_begin
     - добавлен: test-has_end

~~9) добавлен: test-200-types/types/void_t~~
~~   - добавлен: test-void_t-new~~
~~   - добавлен: test-void_t-old~~
     - закомментирован, потому что не удалось заставить работать

**Версия 1.2.5 (предварительная)**  
----------------------------------
[2021y-04m-12d][01:39:34] #59_docs_types
1) рефактор документации:

include/tools/types/types.hpp
// [2021y-03m-16d][16:55:29] Idrisov Denis R. 100
// [2021y-03m-17d][23:51:45] Idrisov Denis R. 101
// [2021y-03m-19d][19:00:21] Idrisov Denis R. 102
// [2021y-03m-19d][22:55:07] Idrisov Denis R. 103
// [2021y-03m-20d][05:33:09] Idrisov Denis R. 104
// [2021y-03m-20d][19:42:05] Idrisov Denis R. 105
// [2021y-03m-21d][17:51:39] Idrisov Denis R. 106
// [2021y-03m-22d][11:09:15] Idrisov Denis R. 107
// [2021y-03m-23d][23:55:08] Idrisov Denis R. 108
// [2021y-03m-24d][21:32:12] Idrisov Denis R. 109
// [2021y-03m-24d][23:57:55] Idrisov Denis R. 110
// [2021y-03m-26d][08:27:10] Idrisov Denis R. 111
// [2021y-03m-27d][23:59:59] Idrisov Denis R. 112
// [2021y-03m-30d][03:40:56] Idrisov Denis R. 113
// [2021y-03m-31d][02:00:09] Idrisov Denis R. 114
// [2021y-03m-31d][22:57:34] Idrisov Denis R. 115
// [2021y-04m-02d][00:40:10] Idrisov Denis R. 116
// [2021y-04m-03d][00:13:46] Idrisov Denis R. 117
// [2021y-04m-03d][23:10:33] Idrisov Denis R. 118
// [2021y-04m-04d][07:05:55] Idrisov Denis R. 119
// [2021y-04m-05d][04:42:08] Idrisov Denis R. 120
// [2021y-04m-07d][00:12:47] Idrisov Denis R. 121
// [2021y-04m-08d][01:36:55] Idrisov Denis R. 122
// [2021y-04m-09d][23:13:29] Idrisov Denis R. 123
// [2021y-04m-09d][23:48:34] Idrisov Denis R. 124
// [2021y-04m-10d][07:15:57] Idrisov Denis R. 125
// [2021y-04m-14d][01:37:05] Idrisov Denis R. 126
// [2021y-04m-19d][14:15:59] Idrisov Denis R. 127 PRE
#pragma once
#ifndef dTOOLS_TYPES_USED_ 
#define dTOOLS_TYPES_USED_ 127

#include <tools/features.hpp>
#include <tools/types/fixed.hpp>
#include <tools/types/common.hpp>
#include <tools/types/variadic.hpp>
#include <tools/types/demangle.hpp>

//==============================================================================
//==============================================================================
#endif // !dTOOLS_TYPES_USED_
